client/network-gossip/src/bridge.rs

```diff
@@ -0,0 +1,299 @@
+	}
+
+	/// Get data of valid, incoming messages for a topic (but might have expired meanwhile)
+	pub fn messages_for(&self, topic: B::Hash)
+		-> mpsc::UnboundedReceiver<TopicNotification>
+	{
+		self.inner.lock().state_machine.messages_for(self.engine_id, topic)
+	}
+
+	/// Send all messages with given topic to a peer.
+	pub fn send_topic(
+		&self,
+		who: &PeerId,
+		topic: B::Hash,
+		force: bool
+	) {
+		let mut inner = self.inner.lock();
+		let inner = &mut *inner;
+		inner.state_machine.send_topic(&mut *inner.context, who, topic, self.engine_id, force)
+	}
+
+	/// Multicast a message to all peers.
+	pub fn gossip_message(
+		&self,
+		topic: B::Hash,
+		message: Vec<u8>,
+		force: bool,
+	) {
+		let message = ConsensusMessage {
+			engine_id: self.engine_id,
+			data: message,
+		};
+
+		let mut inner = self.inner.lock();
+		let inner = &mut *inner;
+		inner.state_machine.multicast(&mut *inner.context, topic, message, force)
+	}
+
+	/// Send addressed message to the given peers. The message is not kept or multicast
+	/// later on.
+	pub fn send_message(&self, who: Vec<network::PeerId>, data: Vec<u8>) {
+		let mut inner = self.inner.lock();
+		let inner = &mut *inner;
+
+		for who in &who {
+			inner.state_machine.send_message(&mut *inner.context, who, ConsensusMessage {
+				engine_id: self.engine_id,
+				data: data.clone(),
+			});
+		}
+	}
+
+	/// Notify everyone we're connected to that we have the given block.
+	///
+	/// Note: this method isn't strictly related to gossiping and should eventually be moved
+	/// somewhere else.
+	pub fn announce(&self, block: B::Hash, associated_data: Vec<u8>) {
+		self.inner.lock().context_ext.announce(block, associated_data);
+	}
+
+	/// Notifies the sync service to try and sync the given block from the given
+	/// peers.
+	///
+	/// If the given vector of peers is empty then the underlying implementation
+	/// should make a best effort to fetch the block from any peers it is
+	/// connected to (NOTE: this assumption will change in the future #3629).
+	///
+	/// Note: this method isn't strictly related to gossiping and should eventually be moved
+	/// somewhere else.
+	pub fn set_sync_fork_request(&self, peers: Vec<network::PeerId>, hash: B::Hash, number: NumberFor<B>) {
+		self.inner.lock().context_ext.set_sync_fork_request(peers, hash, number);
+	}
+}
+
+impl<B: BlockT> Clone for GossipEngine<B> {
+	fn clone(&self) -> Self {
+		GossipEngine {
+			inner: self.inner.clone(),
+			engine_id: self.engine_id.clone(),
+		}
+	}
+}
+
+struct ContextOverService<N> {
+	network: N,
+}
+
+impl<B: BlockT, N: Network<B>> Context<B> for ContextOverService<N> {
+	fn report_peer(&mut self, who: PeerId, reputation: ReputationChange) {
+		self.network.report_peer(who, reputation);
+	}
+
+	fn disconnect_peer(&mut self, who: PeerId) {
+		self.network.disconnect_peer(who)
+	}
+
+	fn send_consensus(&mut self, who: PeerId, messages: Vec<ConsensusMessage>) {
+		// TODO: send batch
```

Comment:
can't merge TODO

Reply:
There's now a `ConsensusBatch` message that groups multiple consensus messages into one, but I don't really see what this brings and looks like an XY solution to me.
This TODO is here because ideally we should send a `ConsensusBatch` message, but I also don't think there's a point in doing so.

Reply:
If we split the gossiping into multiple substreams (as planned), then the `ConsensusBatch` will disappear anyway.

Reply:
> This TODO is here because ideally we should send a ConsensusBatch message, but I also don't think there's a point in doing so.

That's fine, but it needs an issue filed if you want this PR to be merged